File: code/DFS/40. 组合总和 II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        '''
        难点：1、数字不重复 2、和为target
        '''
        ans = []
        candidates.sort() # 从小到大排序
        def dfs(curIndex, curList):
            curTotal = sum(curList)
            if curTotal > target:
                return
            if curTotal == target:
                ans.append(curList)
                return
            i = curIndex
            while i < len(candidates):
                dfs(i+1, curList+[candidates[i]])
                # 避免出现重复的方案，所以一直往后找到和当前数字不一样的
                while i+1 < len(candidates) and candidates[i] == candidates[i+1]:
                    i+=1
                i += 1

        dfs(0, [])
        return ans
            

File: code/递归和回溯/40. 组合总和 II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates = candidates.sort((a, b)=>(a-b)) // 从小到大排序
    const ans = []
    dfs(0, [], target, ans, candidates)
    return ans
};
const sum = lst => lst.reduce((preValue, curValue, curIndex, array)=> preValue+curValue)
function dfs(curIndex, curList, target, ans, candidates){
    const curTotal = curList.length ? sum(curList) : 0
    if (curTotal > target){
        return
    }
    if(curTotal == target){
        ans.push([...curList])
        return
    }
    let i = curIndex
    while(i < candidates.length){
        dfs(i+1, [...curList, candidates[i]], target, ans, candidates)
        while(i+1 < candidates.length && candidates[i] == candidates[i+1]){
            i++
        }
        i++
    }

}